plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/expression/OperatorExpEmbeddings.kt

```diff
@@ -0,0 +1,232 @@
+        viperImplementation { Exp.Or(args[0], args[1], pos, info, trafos) }
+    }
+
+    val Implies = buildOperatorExpEmbedding {
+        setName("impliesBools")
+        setSignature(booleanBooleanToBooleanType)
+        viperImplementation { Exp.Implies(args[0], args[1], pos, info, trafos) }
+    }
+
+    val SubCharChar = buildOperatorExpEmbedding {
+        setName("subChars")
+        withSignature {
+            withParam { char() }
+            withParam { char() }
+            withReturnType { int() }
+        }
+        viperImplementation { Exp.Sub(args[0], args[1], pos, info, trafos) }
+    }
+
+    private val charIntToCharType = buildFunctionPretype {
+        withParam { char() }
+        withParam { int() }
+        withReturnType { char() }
+    }
+
+    val AddCharInt = buildOperatorExpEmbedding {
+        setName("addCharInt")
+        setSignature(charIntToCharType)
+        viperImplementation { Exp.Add(args[0], args[1], pos, info, trafos) }
+    }
+
+    val SubCharInt = buildOperatorExpEmbedding {
+        setName("subCharInt")
+        setSignature(charIntToCharType)
+        viperImplementation { Exp.Sub(args[0], args[1], pos, info, trafos) }
+    }
+
+    private val charCharToBooleanType = buildFunctionPretype {
+        withParam { char() }
+        withParam { char() }
+        withReturnType { boolean() }
+    }
+
+    val GeCharChar = buildOperatorExpEmbedding {
+        setName("geChars")
+        setSignature(charCharToBooleanType)
+        viperImplementation { Exp.GeCmp(args[0], args[1], pos, info, trafos) }
+    }
+
+    val GtCharChar = buildOperatorExpEmbedding {
+        setName("gtChars")
+        setSignature(charCharToBooleanType)
+        viperImplementation { Exp.GtCmp(args[0], args[1], pos, info, trafos) }
+    }
+
+    val LeCharChar = buildOperatorExpEmbedding {
+        setName("leChars")
+        setSignature(charCharToBooleanType)
+        viperImplementation { Exp.LeCmp(args[0], args[1], pos, info, trafos) }
+    }
+
+    val LtCharChar = buildOperatorExpEmbedding {
+        setName("ltChars")
+        setSignature(charCharToBooleanType)
+        viperImplementation { Exp.LtCmp(args[0], args[1], pos, info, trafos) }
+    }
+
+    val StringLength = buildOperatorExpEmbedding {
+        setName("stringLength")
+        withSignature {
+            withParam { string() }
+            withReturnType { int() }
+        }
+        viperImplementation { Exp.SeqLength(args[0], pos, info, trafos) }
+    }
+
+    val StringGet = buildOperatorExpEmbedding {
+        setName("stringGet")
+        withSignature {
+            withParam { string() }
+            withParam { int() }
+            withReturnType { char() }
+        }
+        viperImplementation { Exp.SeqIndex(args[0], args[1], pos, info, trafos) }
+        additionalConditions {
+            precondition {
+                listOf(
+                    intInjection.fromRef(args[1]) ge 0.toExp(),
```

Comment:
Why use ge and lt here instead of >= and <=?

Reply:
Oh, it's because `>=` is a call to `compareTo` which should return an `Int` (and then it is transformed to `Boolean`). Here it's just a nicer way to write an `Exp`.